controllers/openstackservergroup_controller.go

```diff
@@ -0,0 +1,214 @@
+						newServerGroup.ObjectMeta.ResourceVersion = ""
+						return !reflect.DeepEqual(oldServerGroup, newServerGroup)
+					},
+				},
+			),
+		).
+		WithEventFilter(predicates.ResourceNotPausedAndHasFilterLabel(ctrl.LoggerFrom(ctx), r.WatchFilterValue)).
+		Complete(r)
+}
+
+func (r *OpenStackServerGroupReconciler) reconcileDelete(scope scope.Scope, openStackServerGroup *infrav1.OpenStackServerGroup) (result ctrl.Result, reterr error) {
+	scope.Logger().Info("Reconciling ServerGroup delete")
+
+	serverGroupName := openStackServerGroup.Name
+
+	computeService, err := compute.NewService(scope)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Get the servergroup by name, even if our K8s resource already has the ID field set.
+	// TODO(dalees): If this returns a 404 do we try to delete with existing UUID? Do we just assume success?
+	serverGroup, err := computeService.GetServerGroupByName(serverGroupName, true)
+	// Retry if the failure was anything other than Not Found.
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if serverGroup != nil {
+		err = computeService.DeleteServerGroup(serverGroup.ID)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	controllerutil.RemoveFinalizer(openStackServerGroup, infrav1.ServerGroupFinalizer)
+	scope.Logger().Info("Reconciled ServerGroup delete successfully")
+	return ctrl.Result{}, nil
+}
+
+func (r *OpenStackServerGroupReconciler) reconcileNormal(ctx context.Context, scope scope.Scope, openStackServerGroup *infrav1.OpenStackServerGroup) (result ctrl.Result, reterr error) {
+
+	// If the OpenStackServerGroup doesn't have our finalizer, add it.
+	if controllerutil.AddFinalizer(openStackServerGroup, infrav1.ServerGroupFinalizer) {
+		scope.Logger().Info("Reconciling ServerGroup: Adding finalizer")
+		// Register the finalizer immediately to avoid orphaning OpenStack resources on delete
+		// NOTE(dalees): This return without Requeue relies on patchServerGroup to persist the change, and the watch triggers an immediate reconcile.
+		return ctrl.Result{}, nil
+	}
+
+	scope.Logger().Info("Reconciling ServerGroup")
+
+	computeService, err := compute.NewService(scope)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	serverGroupStatus, err := r.getOrCreate(scope.Logger(), openStackServerGroup, computeService)
+	if err != nil || serverGroupStatus == nil {
+		return ctrl.Result{}, err
+	}
+
+	// Update the resource with any new information.
+	openStackServerGroup.Status.Ready = true
+	openStackServerGroup.Status.ID = serverGroupStatus.ID
+
+	scope.Logger().Info("Reconciled ServerGroup successfully")
+	return ctrl.Result{}, nil
+}
+
+func (r *OpenStackServerGroupReconciler) getOrCreate(logger logr.Logger, openStackServerGroup *infrav1.OpenStackServerGroup, computeService *compute.Service) (*servergroups.ServerGroup, error) {
+
+	serverGroupName := openStackServerGroup.Name
+
+	serverGroup, err := computeService.GetServerGroupByName(serverGroupName, false)
```

Comment:
Again, we should probably lookup by ID first in case we have duplicate names.

Reply:
agree
instead of `GetServerGroupByName` raising error in case of multiple server groups with the same name, i suggest to check `openStackServerGroup.Status.ID`, if it is nil, create a new server group